internal/bguploader.go

```diff
@@ -1,146 +1,171 @@
+func NewBgUploader(walFilePath string, maxParallelWorkers int32, maxNumUploaded int32, uploader *WalUploader, preventWalOverwrite bool) *BgUploader {
+	started := make(map[string]struct{})
+	started[filepath.Base(walFilePath)+readySuffix] = struct{}{}
+	ctx, cancelFunc := context.WithCancel(context.Background())
 	return &BgUploader{
-		filepath.Dir(walFilePath),
-		0,
-		maxParallelWorkers,
-		sync.WaitGroup{},
-		started,
-		uploader,
-		0,
-		sync.Mutex{},
-		preventWalOverwrite,
+		dir:                 filepath.Dir(walFilePath),
+		uploader:            uploader,
+		preventWalOverwrite: preventWalOverwrite,
+
+		ctx:                ctx,
+		cancelFunc:         cancelFunc,
+		workerCount:        semaphore.NewWeighted(int64(maxParallelWorkers)),
+		maxParallelWorkers: maxParallelWorkers,
+		numUploaded:        0,
+		maxNumUploaded:     maxNumUploaded,
+		started:            started,
 	}
 }
 
 // Start up checking what's inside archive_status
-func (bgUploader *BgUploader) Start() {
-	if bgUploader.maxParallelWorkers < 1 {
-		return // Nothing to start
+func (b *BgUploader) Start() {
+	// Exit early if BgUploader is effectively disabled
+	if b.maxParallelWorkers < 1 || b.maxNumUploaded < 1 {
+		return
 	}
-	// This goroutine will spawn new if necessary
-	go bgUploader.scanOnce()
-}
 
-// Stop pipeline
-func (bgUploader *BgUploader) Stop() {
-	for atomic.LoadInt32(&bgUploader.parallelWorkers) != 0 {
-		time.Sleep(50 * time.Millisecond)
-	} // Wait until no one works
-
-	bgUploader.mutex.Lock()
-	// We have to do this under mutex to exclude interference with shouldKeepScanning() branch
-	atomic.StoreInt32(&bgUploader.maxParallelWorkers, 0) // stop new jobs
-	bgUploader.mutex.Unlock()
+	go b.scanFiles()
+}
 
-	bgUploader.running.Wait() // wait again for those who jumped to the closing door
+// Stop pipeline. Stop can be safely called concurrently and repeatedly.
+func (b *BgUploader) Stop() {
+	// Send signal to stop scanning for and uploading new files
+	b.cancelFunc()
+	// Wait for all running uploads
+	b.workerCount.Acquire(context.TODO(), int64(b.maxParallelWorkers))
 }
 
 var readySuffix = ".ready"
 var archiveStatus = "archive_status"
 
-// TODO : unit tests
-func (bgUploader *BgUploader) scanOnce() {
-	bgUploader.mutex.Lock()
-	defer bgUploader.mutex.Unlock()
+func (b *BgUploader) scanFiles() {
+	fileChan := make(chan os.FileInfo)
+	defer close(fileChan)
+	go b.processFiles(fileChan)
 
-	files, err := ioutil.ReadDir(filepath.Join(bgUploader.dir, archiveStatus))
-	if err != nil {
-		tracelog.ErrorLogger.Print("Error of parallel upload: ", err)
-		return
+	for {
+		files, err := ioutil.ReadDir(filepath.Join(b.dir, archiveStatus))
+		if err != nil {
+			tracelog.ErrorLogger.Print("Error of parallel upload: ", err)
+			return
+		}
+
+		for _, f := range files {
+			select {
+			case <-b.ctx.Done():
+				return
+			case fileChan <- f:
+			}
+		}
+
+		// Sleep 5 seconds before scanning filesystem again. Exit if
```

Comment:
Can you expand why sleep for 5 seconds? Is it to avoid busy-loop scanning the file-system? It's got a bit of a magic constant feel.

Reply:
I agree. Also waiting for 5 seconds seems way too long. Tiplically WALs are uplodad 4-16 files per second.

Reply:
I'll reduce this wait to... 1 second?

In production, we've noticed that when wal-g falls behind and we accumulate 100GB's-1TB's of WAL, `ls pg_wal > /dev/null` takes up to one second. I wanted to add a conservative wait time in order to avoid scanning the filesystem more than strictly necessary. 

Since all of BgUploader's functionality happens while `HandleWALPush` uploads the initial WAL file, I expect 1 scan & process iteration to be sufficient. Example:
- 10 "ready" WAL segments are on disk
- `wal-g wal-push` the first file
- HandleWALPush starts uploading the first file and invokes BgUploader.start()
- BgUploader scans and finds the other 9 files
- BgUploader starts uploading the other 9 files
- HandleWALPush completes uploading the first file and calls BgUploader.Stop()
- wait until BgUploader.Stop() returns, indicating that all 9 files are complete
- exit

Assuming that all files are roughly the same size and therefore take the same amount of time to upload, it shouldn't be necessary or possible for scanAndProcessFiles need to do more than 1 loop (scan the filesystem more than once).

Alternatively, we could try adopting something like https://github.com/fsnotify/fsnotify to avoid scanning the entire directory (probably unnecessary and out of scope for this PR/project).

Reply:
We could reuse open TCP connections to upload WALs if we upload not 9 but 32 files. And avoid wal-g startup costs. And if ls is so long - I think we should memorize results from previous run if it was not 100% used.

Reply:
I've reduced this to 100 milliseconds.